plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/DeployJob.java

```diff
@@ -133,28 +112,54 @@ public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
         return saveStatus;
       }
 
-      IStatus stagingStatus = stageProject(credentialFile, stagingDirectory, progress.newChild(30));
+      synchronized (this) {
```

Comment:
This part may change a bit, but I'll explain anyway. It's guarding the section against `canceling()`. I believe `canceling()` can be executing while the job is running. It's called when a cancellation request is made:

```
	 * A hook method indicating that this job is running and {@link #cancel()}
	 * is being called for the first time.
	 * <p>
	 * Subclasses may override this method to perform additional work when
	 * a cancelation request is made.  This default implementation does nothing.
```